contracts/contracts/UmbraRouter.sol

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+interface IUmbra {
+  function sendEth(
+    address payable receiver,
+    uint256 tollCommitment,
+    bytes32 pkx,
+    bytes32 ciphertext
+  ) external payable;
+
+  function sendToken(
+    address receiver,
+    address tokenAddr,
+    uint256 amount,
+    bytes32 pkx,
+    bytes32 ciphertext
+  ) external payable;
+}
+
+contract UmbraRouter {
+  address internal constant umbra = 0xFb2dc580Eed955B528407b4d36FfaFe3da685401;
+
+  struct SendEth {
+    address payable receiver;
+    uint256 amount;
+    bytes32 pkx;
+    bytes32 ciphertext;
+  }
+
+  struct SendToken {
+    address receiver;
+    address tokenAddr;
+    uint256 amount;
+    bytes32 pkx;
+    bytes32 ciphertext;
+  }
+
+  function batchSendEth(uint256 _tollCommitment, SendEth[] calldata _params) public payable {
+    uint256 balance = msg.value;
+
+    for (uint256 i = 0; i < _params.length; i++) {
+      //amount to be sent per receiver
+      uint256 _amount = _params[i].amount;
+      require(balance >= _amount, "not enough msg.value");
+      balance -= _amount;
+
+      bytes memory data =
+        abi.encodeWithSelector(
+          IUmbra.sendEth.selector,
+          _params[i].receiver,
+          _tollCommitment,
+          _params[i].pkx,
+          _params[i].ciphertext
+        );
+
+      (bool success, ) = umbra.call{value: _amount}(data);
+      require(success, "sendEth call failed");
+    }
+  }
+
+  function batchSendTokens(uint256 _tollCommitment, SendToken[] calldata _params) public payable {
+    for (uint256 i = 0; i < _params.length; i++) {
+      uint256 _amount = _params[i].amount;
+      address _tokenAddr = _params[i].tokenAddr;
+      IERC20 token = IERC20(address(_tokenAddr));
+
+      //User needs to approve router address as spender first
+      token.transferFrom(msg.sender, address(this), _amount);
```

Comment:
I made an attempt at this. The fn is called receiveTokens.

Reply:
This works, but in some cases this will actually be more expensive than the previous approach! If I'm batch sending tokens with one send of DAI, one of USDC, and one of USDT, I still do three transfers, but now I have extra logic and storage writes. And storage refunds are capped at 20% so even though the cost of "[slot started zero, currently nonzero, now being reset to zero](https://github.com/wolflo/evm-opcodes/blob/main/gas.md#a7-sstore)" is only 100 gas because of a 19,900 gas refund, you won't actually realize the full 59700 refund making this new approach a lot pricier.

That's why I think we should leave the original approach for now, and think harder about the options in a follow up PR 🙂

(As a side note, you don't need the nested mapping since the first arg is always `msg.sender` and gets cleared at the end of the transaction, so a standard mapping should work the same)

Reply:
In case you're interested, here are some of the approaches we'll consider:
- Use [memmove](https://github.com/brockelmore/memmove) for in memory arrays and mappings to avoid storage writes (not yet audited). With this approach, we'd basically do what you do now, but in memory, so it's cheaper than storage
- Require the array inputs to be sorted by address. Then we can more easily `require(params[i].tokenAddr >= params[i-1].tokenAddr)` on each loop iteration to simplify tracking the tokens and amounts
- Add a new struct array input called `TransferSummary[]` and have the user tell us what their token sends sum up to, and verify they told us the truth. This is how we did something similar in [dgrants](https://github.com/dcgtc/dgrants/blob/2732f3107fd497b8c7f6ba7a13fcfb9f9f0e89e9/contracts/contracts/GrantRoundManager.sol#L120-L124) but instead of simple token transfers we had varying amounts of tokens to swap